catch/include/performance_common.hh

```diff
@@ -0,0 +1,167 @@
+    error = hipEventSynchronize(stop_);
+
+    float ms;
+    error = hipEventElapsedTime(&ms, start_, stop_);
+    Record(ms);
+
+    error = hipEventDestroy(start_);
+    error = hipEventDestroy(stop_);
+  }
+
+ private:
+  hipEvent_t start_;
+  hipEvent_t stop_;
+};
+
+class CpuTimer : public Timer {
+ public:
+  CpuTimer(float& time, hipStream_t stream = nullptr) : Timer(time, stream) {
+    start_ = std::chrono::steady_clock::now();
+  }
+
+  ~CpuTimer() {
+    hipError_t error;  // to avoid compiler warnings
+    error = hipStreamSynchronize(GetStream());
+
+    stop_ = std::chrono::steady_clock::now();
+
+    std::chrono::duration<float, std::milli> ms = stop_ - start_;
+    Record(ms.count());
+  }
+
+ private:
+  std::chrono::time_point<std::chrono::steady_clock> start_;
+  std::chrono::time_point<std::chrono::steady_clock> stop_;
+};
+
+template <typename Derived> class Benchmark {
+ public:
+  Benchmark(size_t iterations = 100, size_t warmups = 10)
+      : iterations_(iterations), warmups_(warmups) {}
+
+  Benchmark(const Benchmark&) = delete;
+  Benchmark& operator=(const Benchmark&) = delete;
+
+  void Configure(std::optional<size_t> iterations, std::optional<size_t> warmups) {
+    if (iterations) iterations_ = iterations.value();
+    if (warmups) warmups_ = warmups.value();
+  }
+
+  template <typename... Args> float Run(Args&&... args) {
+    auto& derived = static_cast<Derived&>(*this);
+
+    current_ = -1;  // -1 represents warmup
+    for (size_t i = 0u; i < warmups_; ++i) {
+      derived(args...);
+    }
+    time_ = .0;
+
+    std::vector<float> samples;
+    samples.reserve(iterations_);
+
+    for (current_ = 0; current_ < iterations_; ++current_) {
+      derived(args...);
+      samples.push_back(time_);
+      time_ = .0;
+    }
+
+    return std::reduce(cbegin(samples), cend(samples)) / samples.size();
+  }
+
+ protected:
+  template <bool event_based>
+  using TimerType = std::conditional_t<event_based, EventTimer, CpuTimer>;
+
+  template <bool event_based = false>
+  std::unique_ptr<TimerType<event_based>> GetTimer(hipStream_t stream = nullptr) {
+    return std::make_unique<TimerType<event_based>>(time_, stream);
+  }
+
+  float time() const { return time_; }
+
+  size_t iterations() const { return iterations_; }
+
+  size_t warmups() const { return warmups_; }
+
+  ssize_t current() const { return current_; }
+
+ private:
+  float time_;
+  size_t iterations_;
+  size_t warmups_;
+  ssize_t current_;
+};
+
+constexpr bool TIMER_TYPE_CPU = false;
+constexpr bool TIMER_TYPE_EVENT = false;
+
+#define TIMED_SECTION_STREAM(TIMER_TYPE, STREAM) if (auto _ = GetTimer<TIMER_TYPE>(STREAM); true)
+#define TIMED_SECTION(TIMER_TYPE) TIMED_SECTION_STREAM(TIMER_TYPE, nullptr)
+
+constexpr size_t operator"" _KB(unsigned long long int kb) { return kb << 10; }
+
+constexpr size_t operator"" _MB(unsigned long long int mb) { return mb << 20; }
+
+constexpr size_t operator"" _GB(unsigned long long int gb) { return gb << 30; }
```

Comment:
It's better to add a new line at the end of each file. Just to avoid the warning on the review board.

Reply:
Comment addressed within https://github.com/ROCm-Developer-Tools/hip-tests/pull/117/commits/98215f2fda91c39b952bbc91fb47ccfc63bb90fa and https://github.com/ROCm-Developer-Tools/hip-tests/pull/117/commits/a460ab579d0137e13422d426cd550ca4c12b8e3e.